### PR TITLE
Retry on non-200 status codes. Improve resiliency in fetching compute…

### PR DESCRIPTION
**Pull request by austinkelleher**
…r details.

### CODE REVIEW

File: src/util/error.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  cause?: Error;
+  endpoint: string;
+  status: string | number;
+  statusText: string;
+}
+
+/**
+ * This class should be used when a provider API call should not be retried
+ */
+export class IntegrationProviderPermanentAPIError extends IntegrationProviderAPIError {
```
Comment (austinkelleher) [inline src/util/error.ts:17]: We may want to adopt this in the SDK.

File: src/util/error.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { IntegrationProviderAPIError } from '@jupiterone/integration-sdk-core';
+
+interface CreateIntegrationProviderAPIErrorOptions {
```
Comment (ndowmon) [inline src/util/error.ts:3]: Do you need to define this interface, or is there a way to make it inherit from the constructor of `IntegrationProviderAPIError`?

File: src/steps/devices/index.ts [hunk 3]
```diff
@@ -46,13 +47,53 @@ async function iterateMobileDevices(
...
+  }
+
+  if (numComputerDetailFetchFailed) {
+    const errorMessage = `Unable to fetch all computer details (success=${numComputerDetailFetchSuccess}, failed=${numComputerDetailFetchFailed})`;
+
+    logger.publishErrorEvent({
+      name: 'fetch_computer_details_error',
+      message: errorMessage,
+      err: new Error(errorMessage),
+    });
```
Comment (ndowmon) [inline src/steps/devices/index.ts:96]: Does this do some magic to set the step's `StepResultStatus` to `failure`? Just want to be sure that any child steps don't churn data by thinking this step was successful. I'd love an explicit way to set the current step's status.

```typescript

export enum StepResultStatus {
  SUCCESS = 'success',
  FAILURE = 'failure',
  PARTIAL_SUCCESS_DUE_TO_DEPENDENCY_FAILURE = 'partial_success_due_to_dependency_failure',
  DISABLED = 'disabled',
  PENDING_EVALUATION = 'pending_evaluation',
}
```

File: src/jamf/client.ts [hunk 4]
```diff
@@ -39,32 +40,68 @@ interface OnApiRequestErrorParams {
...
+
 interface CreateJamfClientParams {
   host: string;
   username: string;
   password: string;
+  request?: RequestFunction;
   onApiRequestError?: (params: OnApiRequestErrorParams) => void;
 }
 
+const noRetryStatusCodes: number[] = [400, 401, 403, 404, 413];
```
Comment (ndowmon) [inline src/jamf/client.ts:56]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/413
Learn something new every day at J1. Who knew?

**Review by ndowmon — APPROVED** [review]

File: src/util/error.ts (hunk 2 above)
Comment (austinkelleher) [reply src/util/error.ts:3]: Ya this isn't exposed by SDK yet. Going to create an issue in SDK repo for it.

File: src/steps/devices/index.ts (hunk 3 above)
Comment (austinkelleher) [reply src/steps/devices/index.ts]: I simplified this by switching to `IntegrationError`. For awareness, publishing an event that contains the string `error` will mark the entire integration job with `errorsOccurred`

**Review by ndowmon — DISMISSED** [review]

**Review by charlieduong94 — APPROVED** [review]

**Pull request merged**